Configuration errors reset VSAddress to None for all VirtualServers

#### Setup Details
 CIS Version: `2.18.1`      
 Build: `f5networks/k8s-bigip-ctlr:2.18.1`
 BIGIP Version: BIG-IP 17.1.1.3 Build 0.0.5 Point Release 3
 AS3 Version: `3.51.0 build 5`
 Agent Mode: AS3
 Orchestration: K8S
 Orchestration Version: 1.28.13
 Pool Mode: Auto   
 Additional Setup details: Cilium v1.16.0

#### Description

Before we describe our issue please note that we run the `f5-ipam-controller` on version `f5networks/f5-ipam-controller:0.1.11` and all our `VirtualServer` objects are configured with a per-cluster configured `ipamLabel`, the ipam controller are started with these flags:
```
--orchestration=kubernetes
--ipam-provider=f5-ip-provider
--ip-range={"CLUSTER01-IPAM":"10.0.1.10-10.0.1.254"}
--log-level=DEBUG
```

The issue we've observed lately, after upgrading from `2.16.x` to `2.18.x` of the `k8s-bigip-ctlr` is that when we hit a configuration issue of a `VirtualServer` that yields a 422 from the F5 BIG-IP _all_ `VirtualServer` objects move from status `OK` -> `ERROR`, at the same time all VSAddress fields are set to `None`. After a while (seconds) the status changes from `ERROR` -> `OK` but the VSAddress is still set to `None`. After restarting the `k8s-bigip-ctlr` we get all IP addresses back (same as before since they were never changed/removed in the IPAM controller).

At the moment this is causing a lot of other issues for us, please see the Observations sections for relevant logs.

The k8s-bigip-ctlr is started with the following flags:
```
--ingress-class=f5
--credentials-directory /tmp/creds
--bigip-partition=cluster01
--bigip-url=https://lb01.example.com/
--custom-resource-mode=true
--http-client-metrics=true
--insecure=true
--ipam=true
--log-as3-response=false
--log-level=INFO
--manage-ingress=false
--node-label-selector=node-role.kubernetes.io/worker=true
--pool-member-type=auto
```

#### Steps To Reproduce

These steps _should_ be enough to replicate this, given that the rest of the versions and components are started with the correct flags:

1) Create at least two VirtualServers with IPAM label and make sure you get an IP address from the `f5-ipam-controller`.
2) Reconfigure, edit one of the VirtualServers, in a way that is not allowed (from a F5 AS3 perspective AFAICT). In our case we changed the `spec.virtualServerName` from `my-service-vs` to `my-service` this yielded the 422 error code from the F5. The error is fine and that the `VirtualServer` object goes into a error state. What's not good is that the IP address are set to `None` at the same time for the other `VirtualServer` - which is not reflecting the current state. The current state as we've seen is that the IPAM allocated address is still there so the VSAddress field should not be changed?
3) Observe how the status and VSAddress fields changes of the object.
4) Rollback to a version before 2.18, in our case 2.16.1 and perform the same test again.

#### Expected Result 
At most one `VirtualServer`, the one with the configuration error sould be affected, but it should never update/touch any other VirtualServers.

#### Actual Result  
See above.

#### Diagnostic Information
N/A

#### Observations (if any)

One observation we've made are that a lot of logic (code) changed between `2.16` -> `2.18` that is related to the status field of the `VirtualServer`, the important changes that we as users of the `k8s-bigip-ctlr` are these:
* `2.16`: https://github.com/F5Networks/k8s-bigip-ctlr/blob/68eca56971dc872abd8a94f9bb54035936df2d20/config/apis/cis/v1/types.go#L22-L26
* `2.18`: https://github.com/F5Networks/k8s-bigip-ctlr/blob/508f686bf2886a4f95af0f3a822e6ed25bce2ec7/config/apis/cis/v1/types.go#L22-L28

The field names changed and the struct were expanded with more fields. Could something deeper down have changed so that when an error occurs (e.g. 422) all statuses are changed for some reason, not just for that specific object we're trying to change? Maybe something with the `f5-ipam-controller` integration messes things up here?

AFAICT no IPAM addresses were deallocated from the ipam controller during the time we had a problem, they were still present (allocated) but the `k8s-bigip-ctlr` still changed the field to `None` which is wrong.

`f5-ipam-controller` log during the time we reconfigured the VirtualServer:
```
2025/01/21 14:18:39 [DEBUG] Enqueueing on Update: kube-system/cluster01.ipam
2025/01/21 14:18:39 [DEBUG] Processing Key: &{0xc000042160 0xc00045fb80 Update}
2025/01/21 14:18:39 [DEBUG] Updated: kube-system/cluster01.ipam  with Status. Removed
Hostname: my-service.cluster01.k8s.example.com    Key: my-service/my-service.cluster01.k8s.example.com_host    IPAMLabel: CLUSTER01-IPAM    IPAddr: 10.0.1.11   Operation: Delete

2025/01/21 14:18:39 [DEBUG] Enqueueing on Update: kube-system/cluster01.ipam
2025/01/21 14:18:39 [DEBUG] Processing Key: &{0xc00019f340 0xc000042160 Update}
2025/01/21 14:18:40 [DEBUG] Enqueueing on Update: kube-system/cluster01.ipam
2025/01/21 14:18:40 [DEBUG] Processing Key: &{0xc00019f600 0xc00019f340 Update}
2025/01/21 14:18:40 [DEBUG] [CORE] Allocated IP: 10.0.1.11 for Request:
Hostname: my-service.cluster01.k8s.example.com    Key: my-service/my-service.cluster01.k8s.example.com_host    IPAMLabel: CLUSTER01-IPAM    IPAddr:     Operation: Create

2025/01/21 14:18:40 [DEBUG] Updated: kube-system/cluster01.ipam with Status. With IP: 10.0.1.11 for Request:
Hostname: my-service.cluster01.k8s.example.com    Key: my-service/my-service.cluster01.k8s.example.com_host    IPAMLabel: CLUSTER01-IPAM    IPAddr:     Operation: Create

2025/01/21 14:18:40 [DEBUG] Enqueueing on Update: kube-system/cluster01.ipam
2025/01/21 14:18:40 [DEBUG] Processing Key: &{0xc00037a000 0xc00019f600 Update}
```

No other IPAM entries were touched or updated at the same time period AFAICT.

`k8s-bigip-ctlr` log during the time we reconfigured the VirtualServer:
```
2025/01/21 14:18:39 [ERROR] unable to make IPAM Request, will be re-requested soon
E0121 14:18:39.931943       1 worker.go:329] [ERROR] Sync &{my-service VirtualServer my-service 0xc0018ae008 Create  false} failed with unable to make IPAM Request, will be re-requested soon
2025/01/21 14:18:39 [WARNING] Request from cluster local resulted in retry for  CREATE in VIRTUALSERVER my-service/my-service
2025/01/21 14:18:39 [INFO] [Request: 1] cluster local requested CREATE in VIRTUALSERVER my-service/my-service
2025/01/21 14:18:39 [INFO] [Request: 1][AS3] creating a new AS3 manifest
2025/01/21 14:18:40 [INFO] [Request: 1][AS3] posting request to https://lb01.example.com/ for cluster01 tenants
2025/01/21 14:18:40 [ERROR] IP address requested for service: my-service/my-service
2025/01/21 14:18:40 [ERROR] Error while updating VS status:Operation cannot be fulfilled on virtualservers.cis.f5.com "my-service": the object has been modified; please apply your changes to the latest version and try again
2025/01/21 14:18:40 [INFO] [Request: 2] cluster local requested UPDATE in IPAM kube-system/cluster01.ipam
2025/01/21 14:19:09 [ERROR] [Request: 1][AS3] Response from BIG-IP: code: 422 --- tenant:cluster01 --- message: declaration failed
```

--- a/docs/RELEASE-NOTES.rst
+++ b/docs/RELEASE-NOTES.rst
@@ -15,8 +15,8 @@ Bug Fixes
 * `Issue 3679 <https://github.com/F5Networks/k8s-bigip-ctlr/issues/3679>`_: Certificate, CA chain, and private key shown in debug logs
 * `Issue 3655 <https://github.com/F5Networks/k8s-bigip-ctlr/issues/3655>`_: Calico static routes not updated when nodes added/removed from cluster
 * `Issue 3717 <https://github.com/F5Networks/k8s-bigip-ctlr/issues/3717>`_: StaticRoute CNI calico doesn't detect multiple blockaffinities
+* `Issue 3723 <https://github.com/F5Networks/k8s-bigip-ctlr/issues/3723>`_: Configuration errors reset VSAddress to None for all VirtualServers, TransportServers, and IngressLinks
 
-Upgrade notes
 ``````````````
 * For using calico cni with staticRoutingMode, update RBAC permissions to monitor calico blockaffinities resource. See `RBAC <./config_examples/rbac/k8s_rbac.yaml>`_
 

--- a/docs/RELEASE-NOTES.rst
+++ b/docs/RELEASE-NOTES.rst
@@ -17,6 +17,7 @@ Bug Fixes
 * `Issue 3717 <https://github.com/F5Networks/k8s-bigip-ctlr/issues/3717>`_: StaticRoute CNI calico doesn't detect multiple blockaffinities
 * `Issue 3723 <https://github.com/F5Networks/k8s-bigip-ctlr/issues/3723>`_: Configuration errors reset VSAddress to None for all VirtualServers, TransportServers, and IngressLinks
 
+Upgrade notes
 ``````````````
 * For using calico cni with staticRoutingMode, update RBAC permissions to monitor calico blockaffinities resource. See `RBAC <./config_examples/rbac/k8s_rbac.yaml>`_
 

--- a/pkg/controller/controller.go
+++ b/pkg/controller/controller.go
@@ -164,9 +164,7 @@ func NewController(params Params, startController bool) *Controller {
 		manageLoadBalancerClassOnly: params.ManageLoadBalancerClassOnly,
 		clusterRatio:                make(map[string]*int),
 		clusterAdminState:           make(map[string]clustermanager.AdminState),
-		vsAddressesMap:              make(map[string]string),
-		tsAddressesMap:              make(map[string]string),
-		ilAddressesMap:              make(map[string]string),
+		vsAddressesMap:              make(map[resourceRef]string),
 	}
 
 	log.Debug("Controller Created")

--- a/pkg/controller/controller.go
+++ b/pkg/controller/controller.go
@@ -164,6 +164,9 @@ func NewController(params Params, startController bool) *Controller {
 		manageLoadBalancerClassOnly: params.ManageLoadBalancerClassOnly,
 		clusterRatio:                make(map[string]*int),
 		clusterAdminState:           make(map[string]clustermanager.AdminState),
+		vsAddressesMap:              make(map[string]string),
+		tsAddressesMap:              make(map[string]string),
+		ilAddressesMap:              make(map[string]string),
 	}
 
 	log.Debug("Controller Created")

--- a/pkg/controller/controller.go
+++ b/pkg/controller/controller.go
@@ -19,13 +19,14 @@ package controller
 import (
 	"context"
 	"fmt"
-	authv1 "k8s.io/api/authorization/v1"
-	"k8s.io/client-go/dynamic"
 	"net/http"
 	"os"
 	"strings"
 	"time"
 	"unicode"
+
+	authv1 "k8s.io/api/authorization/v1"
+	"k8s.io/client-go/dynamic"
 
 	"github.com/F5Networks/k8s-bigip-ctlr/v2/pkg/vxlan"
 
@@ -164,7 +165,7 @@ func NewController(params Params, startController bool) *Controller {
 		manageLoadBalancerClassOnly: params.ManageLoadBalancerClassOnly,
 		clusterRatio:                make(map[string]*int),
 		clusterAdminState:           make(map[string]clustermanager.AdminState),
-		vsAddressesMap:              make(map[resourceRef]string),
+		ResourceStatusVSAddressMap:  make(map[resourceRef]string),
 	}
 
 	log.Debug("Controller Created")

--- a/pkg/controller/nativeResourceWorker_test.go
+++ b/pkg/controller/nativeResourceWorker_test.go
@@ -193,14 +193,14 @@ var _ = Describe("Routes", func() {
 			mockCtlr.multiClusterHandler.ClusterConfigs[""].namespaces["test"] = struct{}{}
 			rskey := fmt.Sprintf("%v/%v", route1.Namespace, route1.Name)
 			mockCtlr.updateRouteAdmitStatus(rskey, "Route Admitted", "", v1.ConditionTrue)
-			time.Sleep(5 * time.Second)
+			time.Sleep(1 * time.Second)
 			route := mockCtlr.fetchRoute(rskey)
 			Expect(len(route1.Status.Ingress)).To(BeEquivalentTo(1), "Incorrect route admit status")
 			mockCtlr.multiClusterHandler.ClusterConfigs[""].routeClientV1.Routes("default").Create(context.TODO(), route1, metav1.CreateOptions{})
 			mockCtlr.multiClusterHandler.ClusterConfigs[""].routeLabel = " pro in (pro) "
 			mockCtlr.processedHostPath.processedHostPathMap["foo.com/foo"] = route1.ObjectMeta.CreationTimestamp
 			mockCtlr.eraseAllRouteAdmitStatus()
-			time.Sleep(5 * time.Second)
+			time.Sleep(1 * time.Second)
 			route = mockCtlr.fetchRoute(rskey)
 			Expect(len(route.Status.Ingress)).To(BeEquivalentTo(0), "Incorrect route admit status")
 		})

--- a/pkg/controller/nativeResourceWorker_test.go
+++ b/pkg/controller/nativeResourceWorker_test.go
@@ -193,6 +193,7 @@ var _ = Describe("Routes", func() {
 			mockCtlr.multiClusterHandler.ClusterConfigs[""].namespaces["test"] = struct{}{}
 			rskey := fmt.Sprintf("%v/%v", route1.Namespace, route1.Name)
 			mockCtlr.updateRouteAdmitStatus(rskey, "Route Admitted", "", v1.ConditionTrue)
+			time.Sleep(5 * time.Second)
 			route := mockCtlr.fetchRoute(rskey)
 			Expect(len(route1.Status.Ingress)).To(BeEquivalentTo(1), "Incorrect route admit status")
 			mockCtlr.multiClusterHandler.ClusterConfigs[""].routeClientV1.Routes("default").Create(context.TODO(), route1, metav1.CreateOptions{})

--- a/pkg/controller/nativeResourceWorker_test.go
+++ b/pkg/controller/nativeResourceWorker_test.go
@@ -199,7 +199,7 @@ var _ = Describe("Routes", func() {
 			mockCtlr.multiClusterHandler.ClusterConfigs[""].routeLabel = " pro in (pro) "
 			mockCtlr.processedHostPath.processedHostPathMap["foo.com/foo"] = route1.ObjectMeta.CreationTimestamp
 			mockCtlr.eraseAllRouteAdmitStatus()
-			time.Sleep(1 * time.Second)
+			time.Sleep(5 * time.Second)
 			route = mockCtlr.fetchRoute(rskey)
 			Expect(len(route.Status.Ingress)).To(BeEquivalentTo(0), "Incorrect route admit status")
 		})

--- a/pkg/controller/responseHandler.go
+++ b/pkg/controller/responseHandler.go
@@ -81,7 +81,11 @@ func (ctlr *Controller) responseHandler(respChan chan resourceStatusMeta) {
 							ctlr.updateVSStatus(virtual, "", StatusError, errors.New(tenantResponse.message))
 						} else {
 							// update the status for virtual server as tenant posting is success
-							ctlr.updateVSStatus(virtual, ctlr.vsAddressesMap[rscKey], StatusOk, nil)
+							ctlr.updateVSStatus(virtual, ctlr.vsAddressesMap[resourceRef{
+								name: virtual.Name,
+								namespace: virtual.Namespace,
+								kind: VirtualServer,
+							}], StatusOk, nil)
 							// Update Corresponding Service Status of Type LB
 							if !ctlr.isAddingPoolRestricted(ctlr.multiClusterHandler.LocalClusterName) {
 								// set status of all the LB services associated with this VS
@@ -113,7 +117,11 @@ func (ctlr *Controller) responseHandler(respChan chan resourceStatusMeta) {
 							ctlr.updateTSStatus(virtual, "", StatusError, errors.New(tenantResponse.message))
 						} else {
 							// update the status for transport server as tenant posting is success
-							ctlr.updateTSStatus(virtual, ctlr.tsAddressesMap[rscKey], StatusOk, nil)
+							ctlr.updateTSStatus(virtual, ctlr.vsAddressesMap[resourceRef{
+								name: virtual.Name,
+								namespace: virtual.Namespace,
+								kind: TransportServer,
+							}], StatusOk, nil)
 							// set status of all the LB services associated with this TS
 							go ctlr.updateLBServiceStatusForVSorTS(virtual, virtual.Status.VSAddress, true)
 						}
@@ -142,7 +150,11 @@ func (ctlr *Controller) responseHandler(respChan chan resourceStatusMeta) {
 							ctlr.updateILStatus(il, "", StatusError, errors.New(tenantResponse.message))
 						} else {
 							// update the status for ingresslink as tenant posting is success
-							ctlr.updateILStatus(il, ctlr.ilAddressesMap[rscKey], StatusOk, nil)
+							ctlr.updateILStatus(il, ctlr.vsAddressesMap[resourceRef{
+								name: il.Name,
+								namespace: il.Namespace,
+								kind: IngressLink,
+							}], StatusOk, nil)
 						}
 					}
 

--- a/pkg/controller/responseHandler.go
+++ b/pkg/controller/responseHandler.go
@@ -3,10 +3,11 @@ package controller
 import (
 	"container/list"
 	"errors"
-	ficV1 "github.com/F5Networks/f5-ipam-controller/pkg/ipamapis/apis/fic/v1"
-	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
 	"strings"
 	"sync"
+
+	ficV1 "github.com/F5Networks/f5-ipam-controller/pkg/ipamapis/apis/fic/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
 
 	log "github.com/F5Networks/k8s-bigip-ctlr/v2/pkg/vlogger"
 	v1 "k8s.io/api/core/v1"
@@ -82,9 +83,9 @@ func (ctlr *Controller) responseHandler(respChan chan resourceStatusMeta) {
 						} else {
 							// update the status for virtual server as tenant posting is success
 							ctlr.updateVSStatus(virtual, ctlr.vsAddressesMap[resourceRef{
-								name: virtual.Name,
+								name:      virtual.Name,
 								namespace: virtual.Namespace,
-								kind: VirtualServer,
+								kind:      VirtualServer,
 							}], StatusOk, nil)
 							// Update Corresponding Service Status of Type LB
 							if !ctlr.isAddingPoolRestricted(ctlr.multiClusterHandler.LocalClusterName) {
@@ -118,9 +119,9 @@ func (ctlr *Controller) responseHandler(respChan chan resourceStatusMeta) {
 						} else {
 							// update the status for transport server as tenant posting is success
 							ctlr.updateTSStatus(virtual, ctlr.vsAddressesMap[resourceRef{
-								name: virtual.Name,
+								name:      virtual.Name,
 								namespace: virtual.Namespace,
-								kind: TransportServer,
+								kind:      TransportServer,
 							}], StatusOk, nil)
 							// set status of all the LB services associated with this TS
 							go ctlr.updateLBServiceStatusForVSorTS(virtual, virtual.Status.VSAddress, true)
@@ -151,9 +152,9 @@ func (ctlr *Controller) responseHandler(respChan chan resourceStatusMeta) {
 						} else {
 							// update the status for ingresslink as tenant posting is success
 							ctlr.updateILStatus(il, ctlr.vsAddressesMap[resourceRef{
-								name: il.Name,
+								name:      il.Name,
 								namespace: il.Namespace,
-								kind: IngressLink,
+								kind:      IngressLink,
 							}], StatusOk, nil)
 						}
 					}

--- a/pkg/controller/responseHandler.go
+++ b/pkg/controller/responseHandler.go
@@ -82,7 +82,7 @@ func (ctlr *Controller) responseHandler(respChan chan resourceStatusMeta) {
 							ctlr.updateVSStatus(virtual, "", StatusError, errors.New(tenantResponse.message))
 						} else {
 							// update the status for virtual server as tenant posting is success
-							ctlr.updateVSStatus(virtual, ctlr.vsAddressesMap[resourceRef{
+							ctlr.updateVSStatus(virtual, ctlr.ResourceStatusVSAddressMap[resourceRef{
 								name:      virtual.Name,
 								namespace: virtual.Namespace,
 								kind:      VirtualServer,
@@ -118,7 +118,7 @@ func (ctlr *Controller) responseHandler(respChan chan resourceStatusMeta) {
 							ctlr.updateTSStatus(virtual, "", StatusError, errors.New(tenantResponse.message))
 						} else {
 							// update the status for transport server as tenant posting is success
-							ctlr.updateTSStatus(virtual, ctlr.vsAddressesMap[resourceRef{
+							ctlr.updateTSStatus(virtual, ctlr.ResourceStatusVSAddressMap[resourceRef{
 								name:      virtual.Name,
 								namespace: virtual.Namespace,
 								kind:      TransportServer,
@@ -151,7 +151,7 @@ func (ctlr *Controller) responseHandler(respChan chan resourceStatusMeta) {
 							ctlr.updateILStatus(il, "", StatusError, errors.New(tenantResponse.message))
 						} else {
 							// update the status for ingresslink as tenant posting is success
-							ctlr.updateILStatus(il, ctlr.vsAddressesMap[resourceRef{
+							ctlr.updateILStatus(il, ctlr.ResourceStatusVSAddressMap[resourceRef{
 								name:      il.Name,
 								namespace: il.Namespace,
 								kind:      IngressLink,

--- a/pkg/controller/responseHandler.go
+++ b/pkg/controller/responseHandler.go
@@ -81,7 +81,7 @@ func (ctlr *Controller) responseHandler(respChan chan resourceStatusMeta) {
 							ctlr.updateVSStatus(virtual, "", StatusError, errors.New(tenantResponse.message))
 						} else {
 							// update the status for virtual server as tenant posting is success
-							ctlr.updateVSStatus(virtual, virtual.Status.VSAddress, StatusOk, nil)
+							ctlr.updateVSStatus(virtual, ctlr.vsAddressesMap[rscKey], StatusOk, nil)
 							// Update Corresponding Service Status of Type LB
 							if !ctlr.isAddingPoolRestricted(ctlr.multiClusterHandler.LocalClusterName) {
 								// set status of all the LB services associated with this VS
@@ -113,7 +113,7 @@ func (ctlr *Controller) responseHandler(respChan chan resourceStatusMeta) {
 							ctlr.updateTSStatus(virtual, "", StatusError, errors.New(tenantResponse.message))
 						} else {
 							// update the status for transport server as tenant posting is success
-							ctlr.updateTSStatus(virtual, virtual.Status.VSAddress, StatusOk, nil)
+							ctlr.updateTSStatus(virtual, ctlr.tsAddressesMap[rscKey], StatusOk, nil)
 							// set status of all the LB services associated with this TS
 							go ctlr.updateLBServiceStatusForVSorTS(virtual, virtual.Status.VSAddress, true)
 						}
@@ -142,7 +142,7 @@ func (ctlr *Controller) responseHandler(respChan chan resourceStatusMeta) {
 							ctlr.updateILStatus(il, "", StatusError, errors.New(tenantResponse.message))
 						} else {
 							// update the status for ingresslink as tenant posting is success
-							ctlr.updateILStatus(il, il.Status.VSAddress, StatusOk, nil)
+							ctlr.updateILStatus(il, ctlr.ilAddressesMap[rscKey], StatusOk, nil)
 						}
 					}
 

--- a/pkg/controller/types.go
+++ b/pkg/controller/types.go
@@ -88,6 +88,9 @@ type (
 		discoveryMode               discoveryMode
 		clusterRatio                map[string]*int
 		clusterAdminState           map[string]clustermanager.AdminState
+		vsAddressesMap              map[string]string
+		tsAddressesMap              map[string]string
+		ilAddressesMap              map[string]string
 		resourceContext
 	}
 	resourceContext struct {

--- a/pkg/controller/types.go
+++ b/pkg/controller/types.go
@@ -88,9 +88,7 @@ type (
 		discoveryMode               discoveryMode
 		clusterRatio                map[string]*int
 		clusterAdminState           map[string]clustermanager.AdminState
-		vsAddressesMap              map[string]string
-		tsAddressesMap              map[string]string
-		ilAddressesMap              map[string]string
+		vsAddressesMap              map[resourceRef]string
 		resourceContext
 	}
 	resourceContext struct {

--- a/pkg/controller/types.go
+++ b/pkg/controller/types.go
@@ -18,11 +18,12 @@ package controller
 
 import (
 	"container/list"
+	"net/http"
+	"sync"
+
 	"k8s.io/apimachinery/pkg/runtime/schema"
 	"k8s.io/client-go/dynamic"
 	"k8s.io/client-go/informers"
-	"net/http"
-	"sync"
 
 	cisapiv1 "github.com/F5Networks/k8s-bigip-ctlr/v2/config/apis/cis/v1"
 	"github.com/F5Networks/k8s-bigip-ctlr/v2/pkg/vxlan"
@@ -88,7 +89,7 @@ type (
 		discoveryMode               discoveryMode
 		clusterRatio                map[string]*int
 		clusterAdminState           map[string]clustermanager.AdminState
-		vsAddressesMap              map[resourceRef]string
+		ResourceStatusVSAddressMap  map[resourceRef]string
 		resourceContext
 	}
 	resourceContext struct {

--- a/pkg/controller/worker.go
+++ b/pkg/controller/worker.go
@@ -375,6 +375,7 @@ func (ctlr *Controller) processResources() bool {
 		if rKey.event != Create {
 			// update the poolMem cache, clusterSvcResource & resource-svc maps
 			ctlr.deleteResourceExternalClusterSvcRouteReference(rscRefKey)
+			delete(ctlr.vsAddressesMap, rscRefKey.namespace+"/"+rscRefKey.name)
 		}
 
 		err := ctlr.processVirtualServers(virtual, rscDelete)
@@ -474,6 +475,7 @@ func (ctlr *Controller) processResources() bool {
 		if rKey.event != Create {
 			// update the poolMem cache, clusterSvcResource & resource-svc maps
 			ctlr.deleteResourceExternalClusterSvcRouteReference(rscRefKey)
+			delete(ctlr.tsAddressesMap, rscRefKey.namespace+"/"+rscRefKey.name)
 		}
 		err := ctlr.processTransportServers(virtual, rscDelete)
 		if err != nil {
@@ -499,6 +501,7 @@ func (ctlr *Controller) processResources() bool {
 			}
 			// clean the CIS cache
 			ctlr.deleteResourceExternalClusterSvcRouteReference(rsRef)
+			delete(ctlr.ilAddressesMap, rsRef.namespace+"/"+rsRef.name)
 		}
 		err := ctlr.processIngressLink(ingLink, rscDelete)
 		if err != nil {
@@ -1523,6 +1526,7 @@ func (ctlr *Controller) processVirtualServers(
 		for _, vrt := range virtuals {
 			// Updating the virtual server IP Address status for all associated virtuals
 			vrt.Status.VSAddress = ip
+			ctlr.vsAddressesMap[vrt.Namespace+"/"+vrt.Name] = ip
 			passthroughVS := false
 			var tlsProf *cisapiv1.TLSProfile
 			var tlsTermination string
@@ -3164,6 +3168,7 @@ func (ctlr *Controller) processTransportServers(
 	}
 	// Updating the virtual server IP Address status
 	virtual.Status.VSAddress = ip
+	ctlr.tsAddressesMap[virtual.Namespace+"/"+virtual.Name] = ip
 	var rsName string
 	if virtual.Spec.VirtualServerName != "" {
 		rsName = formatCustomVirtualServerName(
@@ -4385,6 +4390,8 @@ func (ctlr *Controller) processIngressLink(
 		if len(ingLink.Spec.IRules) > 0 {
 			rsCfg.Virtual.IRules = ingLink.Spec.IRules
 		}
+		ingLink.Status.VSAddress = ip
+		ctlr.ilAddressesMap[ingLink.Namespace+"/"+ingLink.Name] = ip
 		if ingLink.Spec.BigIPRouteDomain > 0 {
 			if ctlr.PoolMemberType == Cluster {
 				log.Warning("bigipRouteDomain is not supported in Cluster mode")
@@ -4713,51 +4720,6 @@ func getNodeport(svc *v1.Service, servicePort int32) int32 {
 		}
 	}
 	return 0
-}
-
-// Update virtual server status with virtual server address
-func (ctlr *Controller) updateVirtualServerStatus(vs *cisapiv1.VirtualServer, ip string, statusOk string) {
-	// Set the vs status to include the virtual IP address
-	vsStatus := cisapiv1.VirtualServerStatus{VSAddress: ip, Status: statusOk}
-	log.Debugf("Updating VirtualServer Status with %v for resource name:%v , namespace: %v", vsStatus, vs.Name, vs.Namespace)
-	vs.Status = vsStatus
-	vs.Status.VSAddress = ip
-	vs.Status.Status = statusOk
-	config := ctlr.multiClusterHandler.getClusterConfig("")
-	_, updateErr := config.kubeCRClient.CisV1().VirtualServers(vs.ObjectMeta.Namespace).UpdateStatus(context.TODO(), vs, metav1.UpdateOptions{})
-	if nil != updateErr {
-		log.Debugf("Error while updating virtual server status:%v", updateErr)
-		return
-	}
-}
-
-// Update Transport server status with virtual server address
-func (ctlr *Controller) updateTransportServerStatus(ts *cisapiv1.TransportServer, ip string, statusOk string) {
-	// Set the vs status to include the virtual IP address
-	tsStatus := cisapiv1.TransportServerStatus{VSAddress: ip, Status: statusOk}
-	log.Debugf("Updating VirtualServer Status with %v for resource name:%v , namespace: %v", tsStatus, ts.Name, ts.Namespace)
-	ts.Status = tsStatus
-	ts.Status.VSAddress = ip
-	ts.Status.Status = statusOk
-	config := ctlr.multiClusterHandler.getClusterConfig("")
-	_, updateErr := config.kubeCRClient.CisV1().TransportServers(ts.ObjectMeta.Namespace).UpdateStatus(context.TODO(), ts, metav1.UpdateOptions{})
-	if nil != updateErr {
-		log.Debugf("Error while updating Transport server status:%v", updateErr)
-		return
-	}
-}
-
-// Update ingresslink status with virtual server address
-func (ctlr *Controller) updateIngressLinkStatus(il *cisapiv1.IngressLink, ip string) {
-	// Set the vs status to include the virtual IP address
-	ilStatus := cisapiv1.IngressLinkStatus{VSAddress: ip}
-	il.Status = ilStatus
-	config := ctlr.multiClusterHandler.getClusterConfig("")
-	_, updateErr := config.kubeCRClient.CisV1().IngressLinks(il.ObjectMeta.Namespace).UpdateStatus(context.TODO(), il, metav1.UpdateOptions{})
-	if nil != updateErr {
-		log.Debugf("Error while updating ingresslink status:%v", updateErr)
-		return
-	}
 }
 
 // returns service obj with servicename

--- a/pkg/controller/worker.go
+++ b/pkg/controller/worker.go
@@ -375,7 +375,7 @@ func (ctlr *Controller) processResources() bool {
 		if rKey.event != Create {
 			// update the poolMem cache, clusterSvcResource & resource-svc maps
 			ctlr.deleteResourceExternalClusterSvcRouteReference(rscRefKey)
-			delete(ctlr.vsAddressesMap, rscRefKey.namespace+"/"+rscRefKey.name)
+			delete(ctlr.vsAddressesMap, rscRefKey)
 		}
 
 		err := ctlr.processVirtualServers(virtual, rscDelete)
@@ -475,7 +475,7 @@ func (ctlr *Controller) processResources() bool {
 		if rKey.event != Create {
 			// update the poolMem cache, clusterSvcResource & resource-svc maps
 			ctlr.deleteResourceExternalClusterSvcRouteReference(rscRefKey)
-			delete(ctlr.tsAddressesMap, rscRefKey.namespace+"/"+rscRefKey.name)
+			delete(ctlr.vsAddressesMap, rscRefKey)
 		}
 		err := ctlr.processTransportServers(virtual, rscDelete)
 		if err != nil {
@@ -501,7 +501,7 @@ func (ctlr *Controller) processResources() bool {
 			}
 			// clean the CIS cache
 			ctlr.deleteResourceExternalClusterSvcRouteReference(rsRef)
-			delete(ctlr.ilAddressesMap, rsRef.namespace+"/"+rsRef.name)
+			delete(ctlr.vsAddressesMap, rsRef)
 		}
 		err := ctlr.processIngressLink(ingLink, rscDelete)
 		if err != nil {
@@ -1526,7 +1526,11 @@ func (ctlr *Controller) processVirtualServers(
 		for _, vrt := range virtuals {
 			// Updating the virtual server IP Address status for all associated virtuals
 			vrt.Status.VSAddress = ip
-			ctlr.vsAddressesMap[vrt.Namespace+"/"+vrt.Name] = ip
+			ctlr.vsAddressesMap[resourceRef{
+				name:      vrt.Name,
+				namespace: vrt.Namespace,
+				kind:      VirtualServer,
+			}] = ip
 			passthroughVS := false
 			var tlsProf *cisapiv1.TLSProfile
 			var tlsTermination string
@@ -3168,7 +3172,11 @@ func (ctlr *Controller) processTransportServers(
 	}
 	// Updating the virtual server IP Address status
 	virtual.Status.VSAddress = ip
-	ctlr.tsAddressesMap[virtual.Namespace+"/"+virtual.Name] = ip
+	ctlr.vsAddressesMap[resourceRef{
+		name:      virtual.Name,
+		namespace: virtual.Namespace,
+		kind:      TransportServer,
+	}] = ip
 	var rsName string
 	if virtual.Spec.VirtualServerName != "" {
 		rsName = formatCustomVirtualServerName(
@@ -4391,7 +4399,11 @@ func (ctlr *Controller) processIngressLink(
 			rsCfg.Virtual.IRules = ingLink.Spec.IRules
 		}
 		ingLink.Status.VSAddress = ip
-		ctlr.ilAddressesMap[ingLink.Namespace+"/"+ingLink.Name] = ip
+		ctlr.vsAddressesMap[resourceRef{
+			name: ingLink.Name,
+			namespace: ingLink.Namespace,
+			kind: IngressLink,
+		}] = ip
 		if ingLink.Spec.BigIPRouteDomain > 0 {
 			if ctlr.PoolMemberType == Cluster {
 				log.Warning("bigipRouteDomain is not supported in Cluster mode")

--- a/pkg/controller/worker.go
+++ b/pkg/controller/worker.go
@@ -375,7 +375,7 @@ func (ctlr *Controller) processResources() bool {
 		if rKey.event != Create {
 			// update the poolMem cache, clusterSvcResource & resource-svc maps
 			ctlr.deleteResourceExternalClusterSvcRouteReference(rscRefKey)
-			delete(ctlr.vsAddressesMap, rscRefKey)
+			delete(ctlr.ResourceStatusVSAddressMap, rscRefKey)
 		}
 
 		err := ctlr.processVirtualServers(virtual, rscDelete)
@@ -488,7 +488,7 @@ func (ctlr *Controller) processResources() bool {
 		if rKey.event != Create {
 			// update the poolMem cache, clusterSvcResource & resource-svc maps
 			ctlr.deleteResourceExternalClusterSvcRouteReference(rscRefKey)
-			delete(ctlr.vsAddressesMap, rscRefKey)
+			delete(ctlr.ResourceStatusVSAddressMap, rscRefKey)
 		}
 		err := ctlr.processTransportServers(virtual, rscDelete)
 		if err != nil {
@@ -514,7 +514,7 @@ func (ctlr *Controller) processResources() bool {
 			}
 			// clean the CIS cache
 			ctlr.deleteResourceExternalClusterSvcRouteReference(rsRef)
-			delete(ctlr.vsAddressesMap, rsRef)
+			delete(ctlr.ResourceStatusVSAddressMap, rsRef)
 		}
 		err := ctlr.processIngressLink(ingLink, rscDelete)
 		if err != nil {
@@ -1539,7 +1539,7 @@ func (ctlr *Controller) processVirtualServers(
 		for _, vrt := range virtuals {
 			// Updating the virtual server IP Address status for all associated virtuals
 			vrt.Status.VSAddress = ip
-			ctlr.vsAddressesMap[resourceRef{
+			ctlr.ResourceStatusVSAddressMap[resourceRef{
 				name:      vrt.Name,
 				namespace: vrt.Namespace,
 				kind:      VirtualServer,
@@ -3185,7 +3185,7 @@ func (ctlr *Controller) processTransportServers(
 	}
 	// Updating the virtual server IP Address status
 	virtual.Status.VSAddress = ip
-	ctlr.vsAddressesMap[resourceRef{
+	ctlr.ResourceStatusVSAddressMap[resourceRef{
 		name:      virtual.Name,
 		namespace: virtual.Namespace,
 		kind:      TransportServer,
@@ -4412,7 +4412,7 @@ func (ctlr *Controller) processIngressLink(
 			rsCfg.Virtual.IRules = ingLink.Spec.IRules
 		}
 		ingLink.Status.VSAddress = ip
-		ctlr.vsAddressesMap[resourceRef{
+		ctlr.ResourceStatusVSAddressMap[resourceRef{
 			name:      ingLink.Name,
 			namespace: ingLink.Namespace,
 			kind:      IngressLink,

--- a/pkg/controller/worker.go
+++ b/pkg/controller/worker.go
@@ -4413,9 +4413,9 @@ func (ctlr *Controller) processIngressLink(
 		}
 		ingLink.Status.VSAddress = ip
 		ctlr.vsAddressesMap[resourceRef{
-			name: ingLink.Name,
+			name:      ingLink.Name,
 			namespace: ingLink.Namespace,
-			kind: IngressLink,
+			kind:      IngressLink,
 		}] = ip
 		if ingLink.Spec.BigIPRouteDomain > 0 {
 			if ctlr.PoolMemberType == Cluster {

--- a/pkg/controller/worker_test.go
+++ b/pkg/controller/worker_test.go
@@ -134,9 +134,7 @@ var _ = Describe("Worker Tests", func() {
 				options.LabelSelector = mockCtlr.multiClusterHandler.ClusterConfigs[""].nativeResourceSelector.String()
 			},
 		)
-		mockCtlr.vsAddressesMap = make(map[string]string)
-		mockCtlr.tsAddressesMap = make(map[string]string)
-		mockCtlr.ilAddressesMap = make(map[string]string)
+		mockCtlr.vsAddressesMap = make(map[resourceRef]string)
 	})
 
 	Describe("Validating Ingress link functions", func() {
@@ -3972,9 +3970,7 @@ extendedRouteSpec:
 					options.LabelSelector = mockCtlr.multiClusterHandler.ClusterConfigs[""].nativeResourceSelector.String()
 				},
 			)
-			mockCtlr.vsAddressesMap = make(map[string]string)
-			mockCtlr.tsAddressesMap = make(map[string]string)
-			mockCtlr.ilAddressesMap = make(map[string]string)
+			mockCtlr.vsAddressesMap = make(map[resourceRef]string)
 		})
 		It("Virtual Server processing on pod update", func() {
 			mockCtlr.PoolMemberType = NodePortLocal

--- a/pkg/controller/worker_test.go
+++ b/pkg/controller/worker_test.go
@@ -134,6 +134,9 @@ var _ = Describe("Worker Tests", func() {
 				options.LabelSelector = mockCtlr.multiClusterHandler.ClusterConfigs[""].nativeResourceSelector.String()
 			},
 		)
+		mockCtlr.vsAddressesMap = make(map[string]string)
+		mockCtlr.tsAddressesMap = make(map[string]string)
+		mockCtlr.ilAddressesMap = make(map[string]string)
 	})
 
 	Describe("Validating Ingress link functions", func() {
@@ -3969,6 +3972,9 @@ extendedRouteSpec:
 					options.LabelSelector = mockCtlr.multiClusterHandler.ClusterConfigs[""].nativeResourceSelector.String()
 				},
 			)
+			mockCtlr.vsAddressesMap = make(map[string]string)
+			mockCtlr.tsAddressesMap = make(map[string]string)
+			mockCtlr.ilAddressesMap = make(map[string]string)
 		})
 		It("Virtual Server processing on pod update", func() {
 			mockCtlr.PoolMemberType = NodePortLocal

--- a/pkg/controller/worker_test.go
+++ b/pkg/controller/worker_test.go
@@ -134,7 +134,7 @@ var _ = Describe("Worker Tests", func() {
 				options.LabelSelector = mockCtlr.multiClusterHandler.ClusterConfigs[""].nativeResourceSelector.String()
 			},
 		)
-		mockCtlr.vsAddressesMap = make(map[resourceRef]string)
+		mockCtlr.ResourceStatusVSAddressMap = make(map[resourceRef]string)
 	})
 
 	Describe("Validating Ingress link functions", func() {
@@ -3970,7 +3970,7 @@ extendedRouteSpec:
 					options.LabelSelector = mockCtlr.multiClusterHandler.ClusterConfigs[""].nativeResourceSelector.String()
 				},
 			)
-			mockCtlr.vsAddressesMap = make(map[resourceRef]string)
+			mockCtlr.ResourceStatusVSAddressMap = make(map[resourceRef]string)
 		})
 		It("Virtual Server processing on pod update", func() {
 			mockCtlr.PoolMemberType = NodePortLocal

